com.unity.ml-agents.extensions/Runtime/Input/InputActuatorComponent.cs

```diff
@@ -0,0 +1,376 @@
+            uint localId,
+            bool isInHeuristicMode,
+            InputActionAsset asset)
+        {
+            var deviceRequirements = new List<InputControlScheme.DeviceRequirement>
+            {
+                new InputControlScheme.DeviceRequirement
+                {
+                    controlPath = device.path,
+                    isOptional = true,
+                    isAND = false,
+                    isOR = true
+                }
+            };
+
+            if (isInHeuristicMode)
+            {
+                for (var i = 0; i < asset.controlSchemes.Count; i++)
+                {
+                    var scheme = asset.controlSchemes[i];
+                    for (var ii = 0; ii < scheme.deviceRequirements.Count; ii++)
+                    {
+                        deviceRequirements.Add(scheme.deviceRequirements[ii]);
+                    }
+                }
+            }
+
+            var controlSchemeName = mlAgentsControlSchemeName + localId;
+            var inputControlScheme = new InputControlScheme(
+                controlSchemeName,
+                deviceRequirements);
+
+            return inputControlScheme;
+        }
+
+#pragma warning disable 672
+        /// <inheritdoc cref="ActuatorComponent.CreateActuator"/>
+        public override IActuator CreateActuator() { return null; }
+#pragma warning restore 672
+
+        /// <inheritdoc cref="IActuator.ActionSpec"/>
+        public override ActionSpec ActionSpec => m_ActionSpec;
+
+        /// <summary>
+        /// Creates a virtual <see cref="InputDevice"/> with a layout that is based on the <see cref="InputAction"/>s
+        /// in the <see cref="InputActionMap"/> specified by <see cref="PlayerInput.defaultActionMap"/>.
+        /// </summary>
+        /// <param name="layoutName">The custom layout name to use for the <see cref="InputDevice"/> to create.</param>
+        /// <param name="interfaceName">The made up interface name we are using for these virtual devices.  Right now
+        /// it is equal to <see cref="mlAgentsDeviceName"/> + <see cref="BehaviorParameters.BehaviorName"/></param>
+        /// <param name="localId"></param>
+        /// <returns></returns>
+        internal static InputDevice CreateDevice(string layoutName, string interfaceName, uint localId)
+        {
+            // See if our device layout was already registered.
+            var existingLayout = InputSystem.TryFindMatchingLayout(new InputDeviceDescription
+            {
+                interfaceName = interfaceName
+            });
+
+            // Load the device layout based on the controls we created previously.
+            if (string.IsNullOrEmpty(existingLayout))
+            {
+                InputSystem.RegisterLayoutMatcher(layoutName, new InputDeviceMatcher()
+                    .WithInterface($"^({interfaceName}[0-9]*)"));
+            }
+
+            // Actually create the device instance.
+            var device = InputSystem.AddDevice(
```

Comment:
For this use case, probably not so useful to go through the full matching process. Seems like it would make more sense here to simply do `InputSystem.AddDevice(layoutName);` and not otherwise deal with layouts at all.